packages/sdk/src/controllers/SubscriberController.ts

```diff
@@ -115,14 +107,7 @@ export class SubscriberController {
      * @param authRefreshRequest Stringified object of AuthRefreshRequest
      */
     onAuthExpired = async (authRefreshRequest: string) => {
```

Comment:
Maybe that one should be outside of the `subscriberController` concept as it does not follow the usual flow.

In this case, the engine is actually explictly **awaiting** for the handler on the integration to return something.
This is also why it's returning `null` by default: if no handler exists, then the engine will react on the returned `null` to unblock itself. If there is one, the integration will process the authentication and return a credentials object so that the engine can proceed to refetch the assets.

With the new refactoring there is none of this logic, there is probably a way to make it work but in the end that one subscriber (also with `onViewportRequested`) should actually exceptionally expect only one handle.

Reply:
The logic is actually moved to https://github.com/chili-publish/studio-sdk/pull/482/files#diff-8cc56e41f48f0a52e82e4c279dba504747c37e1c76bf0d79b802505878eb56a7R46-R61. What I did was just executing all subscribers, and the return result of the latest subscriber will be used to return to engine. This still allows to override the `onViewportRequested` or `onAuthExpired`. I do agree it might be a bit obscure if we dont document this very well.

Alternatively we indeed need to make special cases for these where we only allow 1 subscription and do not allow overriding that. All in all I dont think we need to make the exception. People adding subscriptions to those methods probably know what they are doing.

Reply:
All the "events" should have a return type of `void`. All the other ones are technically not "events", they are requests that require a singular handler. 

We should distinguish between them. 

----

Perhaps `clone.events` and `clone.requests` 🤔 

See: https://github.com/chili-publish/studio-sdk/pull/482/files#diff-8cc56e41f48f0a52e82e4c279dba504747c37e1c76bf0d79b802505878eb56a7R75-R149

Reply:
Good points.

* I introduced `handlers` next to `events`: https://github.com/chili-publish/studio-sdk/pull/482/files#diff-a5080035ef7a9f26c66428b08da906fd0d75994ea367df52cae7e305db2c33d4R36-R39
* They are managed by a `SingleEventSubscription` which does not allow adding additional handlers.

Reply:
Thanks! One note on the new handlers section.

We used to have what could be a default handler that was returning `null`, the reasoning was to gracefully fail and/or to return a meaningful error message when an handler wasn't subscribed to.

Without those default handlers returning `null`, engine will await indefinitely until something is returned to it and will just hang without any hint.
e.g. on `main`, behavior is that if no viewport is returned, engine will catch the `null`, stop awaiting and return a message saying that the viewport handler needs to be defined.

Unless `trigger` can return `null` by default, or maybe just enforcing the handler to be defined or any other mechanism.

Reply:
I thought it was handled, so added an explicit test case, and you are right, it was returning undefined instead of null in this case. Should be good now!